src/hotspot/share/runtime/synchronizer.cpp

```diff
@@ -1533,17 +1546,32 @@ size_t ObjectSynchronizer::deflate_idle_monitors(ObjectMonitorsHashtable* table)
     }
 
     // After the handshake, safely free the ObjectMonitors that were
-    // deflated in this cycle.
-    for (ObjectMonitor* monitor: delete_list) {
-      delete monitor;
-      deleted_count++;
-
-      if (current->is_Java_thread()) {
-        // A JavaThread must check for a safepoint/handshake and honor it.
-        chk_for_block_req(JavaThread::cast(current), "deletion", "deleted_count",
-                          deleted_count, ls, &timer);
+    // deflated and unlinked in this cycle.
+    if (current->is_Java_thread()) {
+      if (ls != NULL) {
+        timer.stop();
+        ls->print_cr("before setting blocked: unlinked_count=" SIZE_FORMAT
+                     ", in_use_list stats: ceiling=" SIZE_FORMAT ", count="
+                     SIZE_FORMAT ", max=" SIZE_FORMAT,
+                     unlinked_count, in_use_list_ceiling(),
+                     _in_use_list.count(), _in_use_list.max());
+      }
+      // Make the calling JavaThread blocked (safepoint safe) while we
+      // free the ObjectMonitors:
+      ThreadBlockInVM tbivm(JavaThread::cast(current));
+      if (ls != NULL) {
+        ls->print_cr("after setting blocked: in_use_list stats: ceiling="
+                     SIZE_FORMAT ", count=" SIZE_FORMAT ", max=" SIZE_FORMAT,
+                     in_use_list_ceiling(), _in_use_list.count(), _in_use_list.max());
+        timer.start();
       }
+      deleted_count = delete_monitors(delete_list);
+      // ThreadBlockInVM is destroyed here
+    } else {
+      // A non-JavaThread can just free the ObjectMonitors:
+      deleted_count = delete_monitors(delete_list);
```

Comment:
Which non-JavaThread can execute this code?

Reply:
Wasn't it the safepoint cleanup actions where the VMThread could call this?

Reply:
There's a call from the VMThread ObjectSynchronizer::do_final_audit_and_print_stats() now that can call this.

Reply:
The VM Thread can cause a deflation cycle to happen when logging is enabled.

Reply:
The code prior to your change did chk_for_block_req, ie checked for a safepoint.  I was asking above why it did this in this loop since the code just above does a handshake. That is, why would it be necessary? This code is generally run in another thread.
Is the safepoint check the slow part or is the OopStorage release()?  Or both?  Do you have a test to point to that gets behind in freeing ObjectMonitors?

Reply:
Doing a os::free which may call NMT while the thread is in the blocked state is probably okay, but I think you should check this out.  There may be a dcmd that runs at safepoint that does NMT things.

Reply:
@coleenp
> The code prior to your change did chk_for_block_req, ie checked for a safepoint. I was asking above why it did this in this loop since the code just above does a handshake. That is, why would it be necessary?

`deflate_monitor_list()` will deflate up to `MonitorDeflationMax` ObjectMonitors.
The default for that value is `1000000`. So when we're processing the "delete_list"
we used to call `chk_for_block_req()` so that we didn't cause too much of a wait
for the next safepoint (or handshake). We did this to avoid impacting time-to-safepoint.

> This code is generally run in another thread.

Yes, in the MonitorDeflation thread. It's a JavaThread so the VM Thread will wait
for it for safepoints.

So @robehn's idea is to split the release apart from the free. We do the release
when we deflate (or when we throw away a newly allocated ObjectMonitor) and
we do the free while blocked. Even less impact on time-to-safepoint.

> Is the safepoint check the slow part or is the OopStorage release()? Or both? Do you have a test to point to that gets behind in freeing ObjectMonitors?

Doing the `chk_for_block_req()` call up to a million times is a bit costly. I don't
know how costly the `OopStorage release()` is, but we moved that for functional
reasons and not for performance reasons. Can't do `OopStorage release()` while
blocked. Doing a `delete/os::free()` can be costly depending on the C-library.
In any case, pushing work from the JavaThread being in VM mode to the JavaThread
being in blocked mode is a "good thing".

> Doing a os::free which may call NMT while the thread is in the blocked state is probably okay, but I think you should check this out. There may be a dcmd that runs at safepoint that does NMT things.

That's an interesting question. I can't imagine that this is the first call to os::free() that
can run in parallel with a safepoint. Normally, I would ask Zhengyu, but he's not working
on this stuff anymore.

For what it's worth, I've run these changes thru several Tier[1-8] test runs without any
signs of issues...

Reply:
Just to make sure I'm understanding things correctly - it's not
OopStorage::release that has thread state requirements, it's
WeakHandle::release. The WeakHandle operation needs to ensure the pointee has
been nulled out (before calling the OopStorage operation), invoking GC
barriers.

If the claim is that OopStorage::release has thread state requirements, I'd
like more info, since that function is intentionally trying to be low on
requirements.

Reply:
> > Doing a os::free which may call NMT while the thread is in the blocked state is probably okay, but I think you should check this out. There may be a dcmd that runs at safepoint that does NMT things.
> 
> That's an interesting question. I can't imagine that this is the first call to os::free() that can run in parallel with a safepoint. Normally, I would ask Zhengyu, but he's not working on this stuff anymore.

That is fine. In summary mode, NMT record_free will just decrease atomic counters. In detail mode, it will look up an entry from the malloc site table and decrease its atomic counters. Malloc site table is lockless. There is nothing that should block, or require safepoint-less-ness.

Reply:
> Just to make sure I'm understanding things correctly - it's not OopStorage::release that has thread state requirements, it's WeakHandle::release. The WeakHandle operation needs to ensure the pointee has been nulled out (before calling the OopStorage operation), invoking GC barriers.
> 
> If the claim is that OopStorage::release has thread state requirements, I'd like more info, since that function is intentionally trying to be low on requirements.

Hi Kim, no it's not about "OopStorage::release", it's about what the GC expect during a safepoint.
AFAICT we should never use any of the "NativeAccess<>::oop_store(_obj, (oop)NULL);" during a safepoint from a safepoint safe thread.
Which in practice means we can never use OopStorage::release in such case, since we always need to clear it first.